cpp/include/cudf/detail/iterator.cuh

```diff
@@ -0,0 +1,112 @@
+ * limitations under the License.
+ */
+
+
+/** --------------------------------------------------------------------------*
+ * @brief provides column input iterator with nulls replaced with a specified value
+ * @file iterator.cuh
+ * 
+ * The column input iterator is designed to be used as an input
+ * iterator for thrust and cub.
+ *
+ * Usage:
+ * auto iter = make_null_replacement_iterator(column, null_value);
+ * 
+ * The column input iterator returns only a scalar value of data at [id] or 
+ * the null_replacement value passed while creating the iterator.
+ * For non-null column, use 
+ * auto iter = column.begin<Element>();
+ * 
+ * -------------------------------------------------------------------------**/
+
+#pragma once
+
+#include <cudf/cudf.h>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <thrust/iterator/iterator_adaptor.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+/** -------------------------------------------------------------------------*
+ * @brief value accessor of column with null bitmask
+ * A unary functor returns scalar value at `id`.
+ * `operator() (cudf::size_type id)` computes `element` and valid flag at `id`
+ *
+ * the return value for element `i` will return `column[i]`
+ * if it is valid, or `null_replacement` if it is null.
+ *
+ * @tparam Element The type of elements in the column
+ * -------------------------------------------------------------------------**/
+template <typename Element>
+struct value_accessor
+{
+  column_device_view const col;         ///< column view of column in device
+  Element const null_replacement{};     ///< value returned when element is null
+
+/** -------------------------------------------------------------------------*
+ * @brief constructor
+ * @param[in] _col column device view of cudf column
+ * @param[in] null_replacement The value to return for null elements
+ * -------------------------------------------------------------------------**/
+  value_accessor(column_device_view const& _col, Element null_val)
+    : col{_col}, null_replacement{null_val}
+  {
+    // verify valid is non-null, otherwise, is_valid() will crash
+    CUDF_EXPECTS(_col.nullable(), "Unexpected non-nullable column.");
+    CUDF_EXPECTS(_col.head() != nullptr, "non-null data is required");
+  }
+
+  CUDA_DEVICE_CALLABLE
+  Element operator()(cudf::size_type i) const {
+    return col.is_valid_nocheck(i) ?
+      col.element<Element>(i) : null_replacement;
+  }
+};
+
+
+/**
+ * @brief Constructs an iterator over a column's values that replaces null
+ * elements with a specified value.
+ *
+ * Dereferencing the returned iterator for element `i` will return `column[i]`
+ * if it is valid, or `null_replacement` if it is null.
```

Comment:
Document that this iterator is only allowed for nullable columns and add a `@throws` to indicate it will throw if the column is not nullable.

Reply:
updated